tests/components/homematicip_cloud/test_init.py

```diff
@@ -169,81 +160,73 @@
-        await async_setup_component(hass, HMIPC_DOMAIN, {HMIPC_DOMAIN: entry_config})
-        is True
-    )
-    assert not hass.services.async_services().get(HMIPC_DOMAIN)
-
-    # add marker, that services are already setup
-    hass.data[HOMEMATICIP_CLOUD_SERVICES] = True
-    await async_setup_services(hass)
+async def test_setup_two_haps_unload_one_by_one(hass):
+    """Test setup two access points and unload one by one and check services."""
 
-    assert not hass.services.async_services().get(HMIPC_DOMAIN)
+    # Setup AP1
+    mock_config = {HMIPC_AUTHTOKEN: "123", HMIPC_HAPID: "ABC123", HMIPC_NAME: "name"}
+    MockConfigEntry(domain=HMIPC_DOMAIN, data=mock_config).add_to_hass(hass)
+    # Setup AP2
+    mock_config2 = {HMIPC_AUTHTOKEN: "123", HMIPC_HAPID: "ABC1234", HMIPC_NAME: "name2"}
+    MockConfigEntry(domain=HMIPC_DOMAIN, data=mock_config2).add_to_hass(hass)
 
+    with patch("homeassistant.components.homematicip_cloud.HomematicipHAP") as mock_hap:
+        instance = mock_hap.return_value
+        instance.async_setup = CoroutineMock(return_value=True)
+        instance.home.id = "1"
+        instance.home.modelType = "mock-type"
+        instance.home.name = "mock-name"
+        instance.home.currentAPVersion = "mock-ap-version"
+        instance.async_reset = CoroutineMock(return_value=True)
 
-async def test_setup_services_and_do_not_unload(hass):
-    """Test setup services and unload services."""
-    entry_config = {
-        CONF_ACCESSPOINT: "ABC123",
-        CONF_AUTHTOKEN: "123",
-        CONF_NAME: "name",
-    }
+        assert await async_setup_component(hass, HMIPC_DOMAIN, {})
 
-    assert (
-        await async_setup_component(hass, HMIPC_DOMAIN, {HMIPC_DOMAIN: entry_config})
-        is True
-    )
-    assert not hass.services.async_services().get(HMIPC_DOMAIN)
+    hmipc_services = hass.services.async_services()[HMIPC_DOMAIN]
+    assert len(hmipc_services) == 8
 
-    await async_setup_services(hass)
+    config_entries = hass.config_entries.async_entries(HMIPC_DOMAIN)
+    assert len(config_entries) == 2
+    # unload the first AP
+    await hass.config_entries.async_unload(config_entries[0].entry_id)
 
     hmipc_services = hass.services.async_services()[HMIPC_DOMAIN]
     assert len(hmipc_services) == 8
 
-    #  hap is not removed from data with 'hass.data[HMIPC_DOMAIN].pop("ABC123")'
-
-    await async_unload_services(hass)
+    # services still exists
     hmipc_services = hass.services.async_services()[HMIPC_DOMAIN]
     assert len(hmipc_services) == 8
 
-    # remove hap from data
-    hass.data[HMIPC_DOMAIN].pop("ABC123")
-    # set marker, that services are not loaded
-    hass.data[HOMEMATICIP_CLOUD_SERVICES] = False
+    # unload the second AP
+    await hass.config_entries.async_unload(config_entries[1].entry_id)
 
+    # Check services are removed
+    assert not hass.services.async_services().get(HMIPC_DOMAIN)
+
+    # try manually unload services
     await async_unload_services(hass)
```

Comment:
We shouldn't need to do this. Shouldn't we just check that the services have been unregistered?

Reply:
Correct, it is not needed in normal operations. It's required to cover this code:
https://github.com/home-assistant/home-assistant/blob/a6eb7767681fdaf4d1fd73b3f466d06d1905499f/homeassistant/components/homematicip_cloud/services.py#L209-L210
This code avoid unregistering of services, if the service marker is no longer set.

Reply:
OK, i removed the service marker. Should be cleaner now.